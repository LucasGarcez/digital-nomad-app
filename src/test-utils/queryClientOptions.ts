import { QueryClientConfig } from "@tanstack/react-query";

export const queryClientOptions: QueryClientConfig = {
  defaultOptions: {
    queries: {
      retry: false,
      gcTime: Infinity,
    },
    mutations: {
      retry: false,
      gcTime: Infinity,
    },
  },
};
