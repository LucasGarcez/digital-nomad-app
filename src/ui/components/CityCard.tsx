import { ImageBackground, ImageBackgroundProps, Pressable } from "react-native";

import { Link } from "expo-router";
import { useAppTheme } from "../theme/useAppTheme";

import { CityPreview } from "@/src/domain/city/City";
import { useCityToggleFavorite } from "@/src/domain/city/operations/useCityToggleFavorite";
import { BlackOpacity } from "./BlackOpacity";
import { Box, TouchableOpacityBox } from "./Box";
import { Icon } from "./Icon";
import { Text } from "./Text";

type CityCardProps = {
  cityPreview: CityPreview;
  style?: ImageBackgroundProps["style"];
};

export function CityCard({ cityPreview, style }: CityCardProps) {
  const { borderRadii } = useAppTheme();
  const { mutate: toggleFavorite } = useCityToggleFavorite();
  return (
    <Link push href={`/city-details/${cityPreview.id}`} asChild>
      <Pressable>
        <ImageBackground
          source={
            typeof cityPreview.coverImage === "number"
              ? cityPreview.coverImage
              : { uri: cityPreview.coverImage }
          }
          style={[{ width: "100%", height: 280 }, style]}
          imageStyle={{ borderRadius: borderRadii.default }}
        >
          <BlackOpacity />
          <Box flex={1} padding="s24" justifyContent="space-between">
            <TouchableOpacityBox
              alignSelf="flex-end"
              onPress={() => {
                toggleFavorite({ cityId: cityPreview.id, isFavorite: false });
              }}
            >
              <Icon
                name={
                  cityPreview.isFavorite ? "Favorite-fill" : "Favorite-outline"
                }
                color={cityPreview.isFavorite ? "primary" : "text"}
              />
            </TouchableOpacityBox>

            <Box>
              <Text variant="title22">{cityPreview.name}</Text>
              <Text variant="text16">{cityPreview.country}</Text>
            </Box>
          </Box>
        </ImageBackground>
      </Pressable>
    </Link>
  );
}
