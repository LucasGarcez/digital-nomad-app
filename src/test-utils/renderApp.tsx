import HomeScreen from "@/app/(protected)/(tabs)";
import TabLayout from "@/app/(protected)/(tabs)/_layout";
import ExploreScreen from "@/app/(protected)/(tabs)/explore";
import ProfileScreen from "@/app/(protected)/(tabs)/profile";
import ProtectedLayout from "@/app/(protected)/_layout";
import CityDetails from "@/app/(protected)/city-details/[id]";
import SignInScreen from "@/app/sign-in";
import SignUpScreen from "@/app/sign-up";
import { renderRouter } from "expo-router/testing-library";
import { AppStack } from "../ui/navigation/AppStack";

import { ThemeProvider } from "@shopify/restyle";
import { AuthContext, AuthProvider } from "../domain/auth/AuthContext";
import { AuthUser } from "../domain/auth/AuthUser";
import { Repositories } from "../domain/Repositories";
import { Toast } from "../infra/feedbackService/adapters/Toast/Toast";
import { ToastFeedback } from "../infra/feedbackService/adapters/Toast/ToastFeedback";
import { FeedbackProvider } from "../infra/feedbackService/FeedbackProvider";
import { InMemoryRepository } from "../infra/repositories/adapters/inMemory";
import { RepositoryProvider } from "../infra/repositories/RepositoryProvider";
import { inMemoryStorage } from "../infra/storage/adapters/InMemoryStorage";
import { StorageProvider } from "../infra/storage/StorageContext";
import theme from "../ui/theme/theme";

import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import clonedeep from "lodash.clonedeep";
import merge from "lodash.merge";
import { queryClientOptions } from "./queryClientOptions";

type DeepPartial<T> = {
  [P in keyof T]?: T[P] extends object ? DeepPartial<T[P]> : T[P];
};

function MockedAuthProvider({ children }: React.PropsWithChildren) {
  const authUser: AuthUser = {
    email: "lucas@coffstack.com",
    id: "1",
    fullname: "Lucas Garcez",
  };

  return (
    <AuthContext.Provider
      value={{
        isReady: true,
        authUser,
        saveAuthUser: async () => {},
        removeAuthUser: async () => {},
      }}
    >
      {children}
    </AuthContext.Provider>
  );
}

export function renderApp(options?: {
  isAuthenticated?: boolean;
  repositories?: DeepPartial<Repositories>;
}) {
  const queryClient = new QueryClient(queryClientOptions);

  const finalRepository: Repositories = merge(
    clonedeep(InMemoryRepository),
    options?.repositories ?? {}
  );

  const FinalAuthProvider = options?.isAuthenticated
    ? MockedAuthProvider
    : AuthProvider;

  function Wrapper({ children }: React.PropsWithChildren) {
    return (
      <QueryClientProvider client={queryClient}>
        <StorageProvider storage={inMemoryStorage}>
          <FinalAuthProvider>
            <FeedbackProvider value={ToastFeedback}>
              <RepositoryProvider value={finalRepository}>
                <ThemeProvider theme={theme}>
                  {children}
                  <Toast />
                </ThemeProvider>
              </RepositoryProvider>
            </FeedbackProvider>
          </FinalAuthProvider>
        </StorageProvider>
      </QueryClientProvider>
    );
  }

  renderRouter(
    {
      _layout: () => <AppStack />,
      "(protected)/_layout": () => <ProtectedLayout />,
      "(protected)/(tabs)/_layout": () => <TabLayout />,
      "(protected)/(tabs)/index": () => <HomeScreen />,
      "(protected)/(tabs)/explore": () => <ExploreScreen />,
      "(protected)/(tabs)/profile": () => <ProfileScreen />,
      "(protected)/city-details/[id]": () => <CityDetails />,
      "sign-in": () => <SignInScreen />,
      "sign-up": () => <SignUpScreen />,
    },
    { wrapper: Wrapper, initialUrl: "/" }
  );
}
