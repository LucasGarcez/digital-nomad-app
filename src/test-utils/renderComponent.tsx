import { ThemeProvider } from "@shopify/restyle";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { render, RenderOptions } from "@testing-library/react-native";
import { ReactElement } from "react";
import theme from "../ui/theme/theme";
import { queryClientOptions } from "./queryClientOptions";

export const AllTheProviders = ({ children }: React.PropsWithChildren) => {
  const client = new QueryClient(queryClientOptions);

  return (
    <QueryClientProvider client={client}>
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    </QueryClientProvider>
  );
};

export const renderComponent = (
  component: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(component, { wrapper: AllTheProviders, ...options });
