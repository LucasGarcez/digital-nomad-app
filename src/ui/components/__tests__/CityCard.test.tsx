import { renderComponent } from "@/src/test-utils/renderComponent";
import { screen } from "@testing-library/react-native";

import { CityCard } from "../CityCard";

describe("<CityCard />", () => {
  it("should display the city country and favorite icon", () => {
    renderComponent(
      <CityCard
        cityPreview={{
          id: "1",
          country: "Brasil",
          coverImage: "fake-url",
          name: "Rio de Janeiro",
        }}
      />
    );

    screen.debug();
    expect(screen.getByText("Brasil")).toBeOnTheScreen();
    expect(screen.getByTestId("Favorite-outline")).toBeOnTheScreen();
  });
});
