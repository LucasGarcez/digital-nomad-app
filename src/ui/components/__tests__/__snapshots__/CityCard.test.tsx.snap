// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CityCard /> should render the component 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  href="/city-details/1"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  role="link"
>
  <View
    accessibilityIgnoresInvertColors={true}
    style={
      [
        {
          "height": 280,
          "width": "100%",
        },
        undefined,
      ]
    }
  >
    <Image
      source={
        {
          "uri": "fake-url",
        }
      }
      style={
        [
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          {
            "height": 280,
            "width": "100%",
          },
          {
            "borderRadius": 16,
          },
        ]
      }
    />
    <View
      style={
        [
          {
            "backgroundColor": "#1B1B1B",
            "height": "100%",
            "opacity": 0.25,
            "position": "absolute",
            "width": "100%",
          },
        ]
      }
    />
    <View
      style={
        [
          {
            "flex": 1,
            "justifyContent": "space-between",
            "padding": 24,
          },
        ]
      }
    >
      <View
        style={
          [
            {
              "alignSelf": "flex-end",
            },
          ]
        }
      >
        <View
          testID="Favorite-outline"
        />
      </View>
      <View
        style={
          [
            {},
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#FFFFFF",
                "fontFamily": "PoppinsSemiBold",
                "fontSize": 22,
                "lineHeight": 24,
              },
            ]
          }
        >
          Bangkok
        </Text>
        <Text
          style={
            [
              {
                "color": "#FFFFFF",
                "fontFamily": "PoppinsRegular",
                "fontSize": 16,
                "lineHeight": 18,
              },
            ]
          }
        >
          Brasil
        </Text>
      </View>
    </View>
  </View>
</View>
`;
